src/adapter/src/coord/sequencer/inner/peek.rs

```diff
@@ -1200,6 +1215,19 @@ impl Coordinator {
             })?;
         };
 
-        Ok(determination)
+        // Collect the read holds we'll pass to the compute controller when executing the peek.
+        // Usually those will be available in `txn_read_holds` (see above), except if we read from
+        // indexes over constant collections.
+        let read_holds = if determination.timestamp_context.timestamp().is_some() {
+            let txn_holds = self
+                .txn_read_holds
+                .get(session.conn_id())
+                .expect("set above");
+            txn_holds.clone_for(source_bundle)
+        } else {
+            self.acquire_read_holds(source_bundle)
+        };
```

Comment:
I don't love this part. Seems like we should also put read holds for indexes on constant views into `txn_read_holds`? I'm not sure how to do that though. When we select from those indexes, `timeline_context` will be `TimestampIndependent` because we don't take indexes into account when determining it. And that leads to `determination.timestamp_context.timestamp()` being `None`, which leads to us dropping the read holds above. I tried putting them into `txn_read_holds` instead, but that makes various tests in `timeline.slt`.